src/shared/hibernate-util.c

```diff
@@ -28,446 +28,414 @@
-        if (!f) {
-                log_debug_errno(errno, "Failed to open /proc/swaps: %m");
-                return errno == ENOENT ? -EOPNOTSUPP : -errno; /* Convert swap not supported to a recognizable error */
+        r = get_block_device_fd(fd, &swap->devno);
+        if (r < 0)
+                return r;
+
+        r = fd_is_fs_type(fd, BTRFS_SUPER_MAGIC);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to check if swap file '%s' is on Btrfs: %m", swap->path);
+        if (r > 0) {
+                r = btrfs_get_file_physical_offset_fd(fd, &offset_raw);
+                if (r < 0)
+                        return r;
+        } else {
+                _cleanup_free_ struct fiemap *fiemap = NULL;
+
+                r = read_fiemap(fd, &fiemap);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to read extent map for swap file '%s': %m", swap->path);
+
+                offset_raw = fiemap->fm_extents[0].fe_physical;
         }
 
+        swap->offset = offset_raw / page_size();
+        return 0;
+}
+
+static int read_swap_entries(SwapEntries *ret) {
+        _cleanup_(swap_entries_done) SwapEntries entries = {};
+        _cleanup_fclose_ FILE *f = NULL;
+
+        assert(ret);
+
+        f = fopen("/proc/swaps", "re");
+        if (!f)
+                return log_debug_errno(errno, "Failed to open /proc/swaps: %m");
+
+        /* Remove header */
         (void) fscanf(f, "%*s %*s %*s %*s %*s\n");
-        for (unsigned i = 1;; i++) {
-                _cleanup_(swap_entry_freep) SwapEntry *swap = NULL;
+
+        for (;;) {
+                _cleanup_(swap_entry_done) SwapEntry swap = {};
                 _cleanup_free_ char *type = NULL;
-                uint64_t swap_offset = 0;
                 int k;
 
-                swap = new(SwapEntry, 1);
-                if (!swap)
-                        return -ENOMEM;
-
-                *swap = (SwapEntry) {
-                        .type = _SWAP_TYPE_INVALID,
-                };
-
                 k = fscanf(f,
                            "%ms "       /* device/file path */
                            "%ms "       /* type of swap */
                            "%" PRIu64   /* swap size */
                            "%" PRIu64   /* used */
-                           "%i\n",      /* priority */
-                           &swap->path, &type, &swap->size, &swap->used, &swap->priority);
+                           "%i"         /* priority */
+                           "\n",
+                           &swap.path, &type, &swap.size, &swap.used, &swap.priority);
                 if (k == EOF)
                         break;
-                if (k != 5) {
-                        log_debug("Failed to parse /proc/swaps:%u, ignoring", i);
-                        continue;
-                }
+                if (k != 5)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to parse /proc/swaps: %m");
```

Comment:
%m shouldn't be used here. Also, the previous message printed the offending line, which was quite useful.